pkg/steps/artifacts.go

```diff
@@ -207,43 +208,66 @@ func (c podClient) WithNewLoggingClient() PodClient {
 	}
 }
 
-func waitForContainer(podClient PodClient, ns, name, containerName string) error {
-	logrus.WithFields(logrus.Fields{
-		"namespace": ns,
-		"name":      name,
-		"container": containerName,
-	}).Trace("Waiting for container to be running.")
-
-	ctx := context.TODO()
+type evaluator func(obj runtime.Object) (bool, error)
 
+// waitForConditionOnObject uses a watch to wait for a condition to be true on an object.
+// When the condition is satisfied or the timeout expires, the object is returned along
+// with any errors encountered.
+func waitForConditionOnObject(ctx context.Context, client watchingclient.Client, identifier ctrlruntimeclient.ObjectKey, list ctrlruntimeclient.ObjectList, into ctrlruntimeclient.Object, evaluate evaluator, timeout time.Duration) error {
 	lw := &cache.ListWatch{
 		ListFunc: func(options metav1.ListOptions) (object runtime.Object, e error) {
-			options.FieldSelector = fields.OneTermEqualSelector("metadata.name", name).String()
-			res := &corev1.PodList{}
-			return res, podClient.List(ctx, res, &ctrlruntimeclient.ListOptions{Namespace: ns, Raw: &options})
+			options.FieldSelector = fields.OneTermEqualSelector("metadata.name", identifier.Name).String()
+			res := list
```

Comment:
Passing in a pointer to the list and dereferencing it to copy in my example was important - I don't know what happens when you don't do that and use the same list object in all areas. cc @alvaroaleman perhaps you have a good suggestion on how to test this?

Reply:
`list ctrlruntimeclient.ObjectList`

It is an interface. So a pointer already?

Reply:
Right, but the _dereference_ is the important part, since it will copy. Re-assigning an interface value to a new variable does not do that.

Reply:
`list *ctrlruntimeclient.ObjectList`: I cannot make the caller to compile if I use it as a type of a param in `waitForConditionOnObject` function.

Reply:
https://github.com/kubernetes-sigs/controller-runtime/pull/1510
Used it for unit test.